src/zigbee2mqtt/zigbee2mqtt-adapter.ts

```diff
@@ -0,0 +1,261 @@
+/**
+ *
+ * ZigbeeClassifier - Determines properties from Zigbee clusters.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.*
+ */
+
+import { Adapter, AddonManagerProxy, Device } from 'gateway-addon';
+import { Config, Zigbee2MQTTAdapter } from '../config';
+import mqtt from 'mqtt';
+import { Zigbee2MqttDevice } from './zigbee2mqtt-device';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const manifest = require('../../manifest.json');
+
+interface Response {
+  data?: {
+    id?: string;
+    block?: boolean;
+    force?: boolean;
+    value: boolean;
+  };
+  status?: string;
+  error?: string;
+}
+
+interface Log {
+  level?: string;
+  message?: string;
+}
+
+const DEVICES_POSTFIX = '/bridge/devices';
+const PERMIT_REQUEST_POSTFIX = '/bridge/request/permit_join';
+const PERMIT_RESPONSE_POSTFIX = '/bridge/response/permit_join';
+const REMOVE_REQUEST_POSTFIX = '/bridge/request/device/remove';
+const REMOVE_RESPONSE_POSTFIX = '/bridge/response/device/remove';
+const LOGGING_POSTFIX = '/bridge/logging';
+
+export class Zigbee2MqttAdapter extends Adapter {
+  private prefix: string;
+
+  private debug: boolean;
+
+  private client?: mqtt.Client;
+
+  constructor(
+    addonManager: AddonManagerProxy,
+    private config: Config,
+    private adapterConfig: Zigbee2MQTTAdapter
+  ) {
+    super(addonManager, manifest.id, manifest.id);
+    this.prefix = adapterConfig.topicPrefix ?? 'zigbee2mqtt';
+    this.debug = config.debugLogs ?? false;
+    this.connect();
+  }
+
+  async connect(): Promise<void> {
+    const host = this.adapterConfig.host ?? 'localhost';
+    const port = this.adapterConfig.port || 1883;
+    const broker = `mqtt://${host}:${port}`;
+    console.log(`Connecting to broker ${broker}`);
+    const client = mqtt.connect(broker);
+    this.client = client;
+
+    client.on('connect', () => {
+      console.log(`Successfully connected to ${broker}`);
+
+      this.subscribe(`${this.prefix}${DEVICES_POSTFIX}`);
+      this.subscribe(`${this.prefix}${PERMIT_RESPONSE_POSTFIX}`);
+      this.subscribe(`${this.prefix}${REMOVE_RESPONSE_POSTFIX}`);
+      if (this.config.zigbee2mqtt?.zigbee2mqttDebugLogs) {
+        this.subscribe(`${this.prefix}${LOGGING_POSTFIX}`);
+      }
+    });
+
+    client.on('error', (error) => {
+      console.error(`Could not connect to broker: ${error}`);
+    });
+
+    client.on('message', (topic, message) => {
+      const raw = message.toString();
+
+      if (this.debug) {
+        console.log(`Received on ${topic}: ${raw}`);
+      }
+
+      try {
+        const json = JSON.parse(raw);
+
+        const parts = topic.split('/');
+
+        if (topic.indexOf(DEVICES_POSTFIX) > -1) {
```

Comment:
Instead of using `.indexOf(...)` for all of these, it would make more sense to just use `topic.endsWith(...)`.

Reply:
done